### PR TITLE
[Yul] SideEffectCollector: take side effect of user defined function into account

### LINKED ISSUE #7078: [Yul] CommonSubexpressionEliminator should take pure-ness of user-defined functions into account
Not only but especially with the introduction of the web assembly glue code, expressions often include user-defined functions. Think of endianness swapping code, for example. Currently, the CSE only considers built-in functions that are pure to not break this identity. We should extend pure-ness to user-defined functions such that CSE/Simplifier will optimize as follows (even without inlining):

```
function add_one(x) -> y { y := add(x, 1) }
let z := add_one(calldataload(0))
let t := add_one(calldataload(0))
let x := sub(z, t)
```
->

```
function add_one(x) -> y { y := add(x, 1) }
let x := 0
```

### PR DESCRIPTION
**Pull request by sifmelcara**
### Description

Fix #7078 ?

@chriseth could you take a look at current progress, to make sure this direction is correct?

What have been done:
1. Extract `SideEffect` as a class to avoid introducing repetitive codes when I expand the functionality of `SideEffectCollector`
2. redesign `SideEffectCollector`. Now we need to give the whole AST to the constructor to construct a `SideEffectCollector`, the constructor will collect side effects information of every function in scope. User can reuse the `SideEffectCollector` object to call `SideEffects sideEffectsOf(Block const& _block)`/`SideEffects sideEffectsOf(Expression const& _expression)`/`SideEffects sideEffectsOf(Statement const& _statement)` to obtain side effect of a piece of code.
3. Now `containsMSize` is no longer purely syntactic. Not sure if this change is desired?
4. Similarly, I'm not sure why `SideEffectCollector` enters into function definitions before? I changed that. Every function definition now have no side effect.

TODO:
1. Implement `CallGraphGenerator`
2. Implement the actual algorithm that determine side effect of each function.
3. Every optimization pass that uses `SideEffectCollector` need to modified. They probably need to construct a single `SideEffectCollector` instance in their constructor and reuse that instance during traversal.

List of file that uses `SideEffectCollector`/`MovableChecker`:
```
libsolidity/analysis/SyntaxChecker.cpp
libyul/optimiser/DataFlowAnalyzer.cpp
libyul/optimiser/ExpressionInliner.cpp
libyul/optimiser/ExpressionSimplifier.cpp
libyul/optimiser/LoadResolver.cpp
libyul/optimiser/RedundantAssignEliminator.cpp
libyul/optimiser/SimplificationRules.cpp
libyul/optimiser/StackCompressor.cpp
libyul/optimiser/UnusedPruner.cpp
```
4. Maybe divide `Semantic.{cpp, h}` into `SideEffectCollector.{cpp, h}` and `TerminationFinder.{cpp, h}`?
5. Tests, tests, tests.

Questions:
1. Is `O((V+E)^2)` algorithm for computing each function's side effects good enough? (`V` and `E` are number of vertex and number of edge in call graph, respectively) If not, we probably need to perform dfs on the condensation graph of the call graph to achieve `O(V+E)`. (SCC can be obtained by `boost/graph/strong_components`)

### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

### CODE REVIEW

**Comment by chriseth** [comment]
I like the fact that you moved the join operator into the side effect class! I'm a little worried that we might confuse one property with another in the constructor calls, but I also don't have a better solution.

Yes, containsMSize is a property that is a little different than the others.

As far as the rest is concerned: I don't think you need to implement the visitors for Blocks and other things. The only information you need is: Which function calls which (including built-in functions). You can ignore FunctionalInstructions, they should not be present anymore in newer code. One you have the call graph, you can start with side effects being associated to the built-in functions and then propagate those accordingly. This should result in at most quadratic runtime in V+E, where V is the functions in the code and not the instructions.

What do you think?

And yes, splitting up the file a little is certainly a good idea.

**Comment by sifmelcara** [comment]
> Yes, containsMSize is a property that is a little different than the others.

So, does containsMSize still need to be purely syntactic? i.e "statement that call an user defined function which contains `msize`" does not `containsMSize`.

> I don't think you need to implement the visitors for Blocks and other things

Not sure if you are talking about `SideEffects sideEffectsOf(Block const& _block)` (because there is no visitor for blocks implemented). This function is for user (other optimization passes) to get side effects of a block.

> The only information you need is: Which function calls which (including built-in functions). You can ignore FunctionalInstructions, they should not be present anymore in newer code. One you have the call graph, you can start with side effects being associated to the built-in functions and then propagate those accordingly.

Is this what you mean?
```
for (YulString b: builtin_functions_with_side_effect)
{
  set<YulString> x = get_all_user_defined_functions_that_directly_or_indirectly_calls_b(b);
  for (YulString f: x)
  {
    sideEffectOfFunction[f] = join(sideEffectOfFunction[f], sideEffectOfFunction[b]);
  }
}
```

**Comment by chriseth** [comment]
`containsMSize` should be a property of the whole program / AST, and not of a function.

**Comment by chriseth** [comment]
I think a single loop would not be enough. Please correct me if I'm wrong, but I think the following is simple, modular and efficient at the same time:

You first build your call graph, ignoring any side-effect business. This can even be a component separate from the side effects collector. This call graph is just a `std::map<YulString, std::set<YulString>>`. It includes both user-defined and built-in functions. Of course, bulit-in functions never appear as keys of the map, only as values.
This call graph only contains direct calls.

In the next stage, you compute the transitive closure of the call graph, i.e. you keep on adding indirect calls until nothing changes.

Finally, you apply the loop you mentioned above: You apply all side-effects of built-in functions to the function that call them. It may be better to first reverse the callgraph with regards to bulit-in functions: Create a map that goes from built-in functions to user-defined functions that call them.

**Comment by sifmelcara** [comment]
> containsMSize should be a property of the whole program / AST, and not of a function.

So is this the reason why `containsMSize` can be purely syntactic?

If yes, can we make `containsMSize` works like other properties, to enable memory-related optimizations in the following program?

```
{
  function f(a) -> x { x := msize() }
  function g(a) -> x { mstore (a, a) }
  g(42)
  let y := mload(42)
}
```

**Comment by chriseth** [comment]
I don't understand. msize is a rather inconvenient opcode that should not be part of code that is about to be optimized. If it is present anywhere in the program, we cannot remove any opcode that touches memory.

**Comment by sifmelcara** [comment]
But in the above program, `f()` is actually never been called, so we can optimize the above program to ` { let y := 42 } `

**Comment by chriseth** [comment]
Sure, but which function calls which is not relevant here. The unused pruner will completely remove `f` and then we can optimize memory, but it is not that a function that does not call `f` can be optimized with regards to memory, because it is a property of the whole program. If `g` does not call `f`, but `f` is called before `g` is called, then we still cannot optimize `g` for memory.

**Comment by sifmelcara** [comment]
> Sure, but which function calls which is not relevant here. The unused pruner will completely remove f and then we can optimize memory, but it is not that a function that does not call f can be optimized with regards to memory, because it is a property of the whole program. If g does not call f, but f is called before g is called, then we still cannot optimize g for memory.

ok! I just want to make sure that I won't break something if it is not "purly syntactic" anymore.

> I think a single loop would not be enough. 

There is `set<YulString> x = get_all_user_defined_functions_that_directly_or_indirectly_calls_b(b);` inside the loop so I think my pseudo code is essentially the same with what you described.

**Comment by chriseth** [comment]
Yep, you are right!

**Comment by sifmelcara** [comment]
Because `SideEffectsCollector` now needs `_ast` as parameter, many optimization class now also need `_ast` (whole yul program) as parameter.

One example is:
`Pattern::matches` need `SideEffectsCollector` https://github.com/ethereum/solidity/blob/48262106626020435051dac51fd2e73f4a80a03f/libyul/optimiser/SimplificationRules.cpp#L190

Thus `SimplificationRules::findFirstMatch` now also needs whole program. And then `KnowledgeBase` and `ExpressionSimplifier`, which needs `findFirstMatch`, also needs whole program as parameter.

Also I'm worried that the optimizer will be slowed down too much. But I don't see a simple way (without refactoring how optimization framework structured) to avoid this.

**Comment by sifmelcara** [comment]
(paste @chriseth 's response here for my future reference)

> chriseth @chriseth 21:32
@sifmelcara there are some places in the code that call `SideEffectsCollector(_dialect, _expr).movable()`
for most of them, we can live with this to fail if the expression is a function call
they usually work on code that has the expression splitter applied which means that the expression is most likely a variable reference or a literal
I might be wrong, but I think it would be good to first check if this would be enough
also, most optimizer runs modify the code in-place. This makes it very fragile to run call analysis on the ast

Ok I can try that. Another option I can think of is to "assume user defined function have worst case scenario side effects if no _ast is given to side effect collector" (which essentially falls back to original side effect collector behavior)

**Comment by chriseth** [comment]
Yep, that sounds good and would essentially fall back to what I proposed.

**Comment by chriseth** [comment]
Another option would be to supply a data structure containing the side-effects of user-defined functions instead of the AST. If a function is not found in that data structure, worst-case side-effects are assumed.

**Comment by chriseth** [comment]
So basically to get the full processing, you would:
 - compute the direct call graph
 - extend it to an indirect call graph
 - extract side-effects of user-defined functions from the indirect call graph 
 - use that structure to compute specific side-effects of snippets of code

**Comment by sifmelcara** [comment]
> You can ignore FunctionalInstructions, they should not be present anymore in newer code.

Hmm, currently I encounter many `FunctionalInstruction` when running tests, not sure how to deal with this.

**Comment by sifmelcara** [comment]
(Current code is still a draft)

Now it somewhat works but there are many cases that user defined function cannot be found (even when running fullSuite), I will investigate that.

Another problem is that there are still FunctionalInstructions present in AST (ex. `ADD` and `OR`).

Edit: And of course I will try to split this into multiple PRs...

File: libsolidity/analysis/SyntaxChecker.cpp [hunk 1]
```diff
@@ -262,10 +262,13 @@ bool SyntaxChecker::visit(InlineAssembly const& _inlineAssembly)
 	if (!m_useYulOptimizer)
 		return false;
 
-	if (yul::SideEffectsCollector(
-		_inlineAssembly.dialect(),
-		_inlineAssembly.operations()
-	).containsMSize())
+	if (
+		// TODO: This should be pure syntatic instead of ignoring code inside function definitions...
```
Comment (chriseth) [inline libsolidity/analysis/SyntaxChecker.cpp:266]: It's probably better to move `containsMSize` to a different class.

Comment (sifmelcara) [reply libsolidity/analysis/SyntaxChecker.cpp:266]: Probably use `ReferencesCounter` to see if `msize` is referenced is enough?

Comment (chriseth) [reply libsolidity/analysis/SyntaxChecker.cpp:266]: Oh yeah, that might work! I'm still a little puzzled why the code still contains FunctionalInstructions.

**Comment by chriseth** [comment]
Will take a look at this.

**Comment by chriseth** [comment]
Replaced by https://github.com/ethereum/solidity/pull/7251

**Pull request closed without merging**